### PR TITLE
[WIP] FW navigation first order hold move to position controller

### PR DESCRIPTION
**Pull request by dagar**
Navigator mission first order hold (FOH) altitude setpoint was limited to non-rotary wing missions within navigator. Given those restrictions it's a much better fit within the fixed wing position controller.

 - fixes https://github.com/PX4/Firmware/issues/6758
 - fixes https://github.com/PX4/Firmware/issues/8852
 - fixes loiter to alt for MC
 - fixes FOH changing altitude during FW loiter within mission

### CODE REVIEW

**Comment by dagar** [comment]
TODO: unify FW position + loitering handling
 - loiter to climb or descend
 - position for large xy

**Comment by dagar** [comment]
@Stifael this kills off the navigator FOH entirely.

The MissionBlock change is rather large. Additional review would be greatly appreciated.

File: src/modules/navigator/mission_block.cpp [hunk 1]
```diff
@@ -116,264 +124,301 @@ MissionBlock::is_mission_item_reached()
...
-					_waypoint_position_reached = true;
+	float dist_xy = -1.0f;
+	float dist_z = -1.0f;
+	const float dist = get_distance_to_point_global_wgs84(lat, lon, alt,
+			   gpos.lat, gpos.lon, gpos.alt, &dist_xy, &dist_z);
 
-					// set required yaw from bearing to the next mission item
-					if (_mission_item.force_heading) {
-						struct position_setpoint_s next_sp = _navigator->get_position_setpoint_triplet()->next;
+	if (!_navigator->get_vstatus()->is_rotary_wing) {
```
Comment (Stifael) [inline src/modules/navigator/mission_block.cpp]: i think it should be without negation

File: src/modules/navigator/mission_block.cpp [hunk 2]
```diff
@@ -116,264 +124,301 @@ MissionBlock::is_mission_item_reached()
...
 
-				if (back_trans_dec > FLT_EPSILON && velocity > FLT_EPSILON) {
-					mission_acceptance_radius = ((velocity / back_trans_dec / 2) * velocity) + reverse_delay * velocity;
+				yaw_required = true;
 
-				}
+				const float yaw_err = _wrap_pi(item.yaw - _navigator->get_local_position()->yaw);
 
+				if (fabsf(yaw_err) > math::radians(_navigator->get_yaw_threshold())) {
+					success = false;
```
Comment (Stifael) [inline src/modules/navigator/mission_block.cpp]: The success and yaw_required boolean are a bit confusing to me. 
Should not first be checked that yaw is required, and if required (https://github.com/PX4/Firmware/pull/8883/files#diff-7a824e3d1e9a5420071c8c9bbe27b8e7R353) it then needs to be checked for success?

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 3]
```diff
@@ -741,18 +741,81 @@ FixedwingPositionControl::control_position(const Vector2f &curr_pos, const Vecto
...
+
+			float dist_xy = -1.0f;
+			float dist_z = -1.0f;
+
+			float dist = get_distance_to_point_global_wgs84(
+					     pos_sp_curr.lat, pos_sp_curr.lon, pos_sp_curr.alt,
+					     _global_pos.lat, _global_pos.lon, _global_pos.alt,
+					     &dist_xy, &dist_z);
+
+			/* close to waypoint, but altitude error greater than twice acceptance */
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: It's a carry over from moving the old code. I use c++ style // for any new comments.

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 4]
```diff
@@ -741,18 +741,81 @@ FixedwingPositionControl::control_position(const Vector2f &curr_pos, const Vecto
...
+			if ((dist >= 0.0f)
+			    && (dist_z > 2 * _parameters.climbout_diff)
+			    && (dist_xy < 2 * pos_sp_curr.acceptance_radius)) {
+
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: Enough distance to cleanly enter the loiter radius. It's a bit arbitrary.

@tstastny might be cleaner to use L1 distance here?

File: src/modules/navigator/mission.cpp [hunk 5]
```diff
@@ -1058,14 +1032,6 @@ Mission::heading_sp_update()
 					_mission_item.yaw = _wrap_pi(yaw + M_PI_F);
 					pos_sp_triplet->current.yaw = _mission_item.yaw;
 
-				} else if (_param_yawmode.get() == MISSION_YAWMODE_FRONT_TO_WAYPOINT
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/navigator/mission.cpp]: I was assuming/anticipating you handling it entirely in the MC position controller. This was done before https://github.com/PX4/Firmware/pull/9146 and needs review.

[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/navigator/mission.cpp]: The ROI portion should have been kept.

File: src/modules/navigator/mission_block.cpp (hunk 1 above)
Comment (dagar) [reply src/modules/navigator/mission_block.cpp]: Yes, good catch.

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 6]
```diff
@@ -741,18 +741,81 @@ FixedwingPositionControl::control_position(const Vector2f &curr_pos, const Vecto
...
+				const float distance_current_previous = get_distance_to_next_waypoint(pos_sp_curr.lat, pos_sp_curr.lon,
+									pos_sp_prev.lat, pos_sp_prev.lon);
+
+				/* Do not try to find a solution if the last waypoint is inside the acceptance radius of the current one */
+				if ((distance_current_previous - pos_sp_curr.acceptance_radius) > FLT_EPSILON) {
+					/* Calculate distance to current waypoint */
+					const float d_current = get_distance_to_next_waypoint(pos_sp_curr.lat, pos_sp_curr.lon, curr_pos(0), curr_pos(1));
+
+					/* Save distance to waypoint if it is the smallest ever achieved, however make sure that
+					 * min_current_sp_distance_xy is never larger than the distance between the current and the previous wp */
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: It's possible for a plane to miss waypoint acceptance and have to turn around and come back.

**Comment by dagar** [comment]
Testing required

 - fixed wing mission testing
  - TAKEOFF
  - WAYPOINT
  - LOITER (timed, unlimited, to alt)
  - test changing the active waypoint (especially back to takeoff, or interrupting a loiter and coming back to it)

 - multicopter
  - LOITER (timed, unlimited, to alt)

**Comment by Antiheavy** [comment]
I just started SITL testing this on FW.  So far haven't noticed anything bad, but I have a lot more testing to do.  Question:  is there such a thing as tangent exit on time based loiters?  I'm guessing not because that wouldn't make sense anyway.
image

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (tstastny) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: yes I would. This is anyway when the controller would start commanding a turn.

**Comment by dagar** [comment]
@Antiheavy thank for the thorough test. We need to think about how the switch between regular waypoint navigation and achieve waypoint altitude via loiter (once arriving) is handled. It might need to latch between one mode or the other a little more carefully.

**Comment by dagar** [comment]
I'd like to pick this up again, starting by reviewing actual FW requirements.

@philipoe

**Comment by dagar** [comment]
@RomanBapst @sfuhrer let's take another look at this one.

**Comment by sfuhrer** [comment]
@dagar I guess we can close this one since https://github.com/PX4/PX4-Autopilot/pull/15677 is in?

**Comment by LorenzMeier** [comment]
Closing as stale.

**Comment by dagar** [comment]
This was finally completed in https://github.com/PX4/PX4-Autopilot/pull/15677.

**Pull request closed without merging**